src/main/java/org/threadly/concurrent/future/WatchdogCache.java

```diff
@@ -0,0 +1,120 @@
+  private final AtomicBoolean cleanerScheduled;
+  
+  /**
+   * Constructs a new {@link WatchdogCache}.  This constructor will use a default static scheduler 
+   * (which is lazily constructed).  This should be fine in most cases, but you can provide your 
+   * own scheduler if you want to avoid the thread creation (which is shared among all instances 
+   * that were constructed with this constructor or {@link Watchdog#Watchdog(long, boolean)}}.
+   * 
+   * @param sendInterruptOnFutureCancel If {@code true}, and a thread is provided with the future, 
+   *                                      an interrupt will be sent on timeout
+   */
+  public WatchdogCache(boolean sendInterruptOnFutureCancel) {
+    this(Watchdog.getStaticScheduler(), sendInterruptOnFutureCancel);
+  }
+
+  /**
+   * Constructs a new {@link WatchdogCache} with a scheduler of your choosing.  It is critical 
+   * that this scheduler has a free thread available to inspect futures which may not have 
+   * completed in the given timeout.  You may want to use a 
+   * {@link org.threadly.concurrent.limiter} to ensure that there are threads available.
+   * 
+   * @param scheduler Scheduler to schedule task to look for expired futures
+   * @param sendInterruptOnFutureCancel If {@code true}, and a thread is provided with the future, 
+   *                                      an interrupt will be sent on timeout
+   */
+  public WatchdogCache(SimpleSchedulerInterface scheduler, boolean sendInterruptOnFutureCancel) {
+    this.scheduler = scheduler;
+    this.sendInterruptOnFutureCancel = sendInterruptOnFutureCancel;
+    cachedDogs = new ConcurrentHashMap<Long, Watchdog>();
+    cacheCleaner = new CleanRunner();
+    cleanerScheduled = new AtomicBoolean(false);
+  }
+  
+  /**
+   * Watch a given {@link ListenableFuture} to ensure that it completes within the provided 
+   * time limit.  If the future is not marked as done by the time limit then it will be 
+   * completed by invoking {@link ListenableFuture#cancel(boolean)}.  Weather a {@code true} or 
+   * {@code false} will be provided to interrupt the running thread is dependent on how this 
+   * {@link WatchdogCache} was constructed.
+   * 
+   * @param future Future to inspect to ensure completion
+   * @param timeoutInMillis Time in milliseconds that future should be completed within
+   */
+  public void watch(ListenableFuture<?> future, long timeoutInMillis) {
+    // attempt around a cheap shortcut
+    if (future == null || future.isDone()) {
+      return;
+    }
+    
+    Watchdog dog = cachedDogs.get(timeoutInMillis);
+    if (dog == null) {
+      dog = new Watchdog(scheduler, timeoutInMillis, sendInterruptOnFutureCancel);
+      Watchdog exisiting = cachedDogs.putIfAbsent(timeoutInMillis, dog);
+      if (exisiting != null) {
+        dog = exisiting;
+      }
+    }
+    
+    dog.watch(future);
+    
+    if (! cleanerScheduled.get() && cleanerScheduled.compareAndSet(false, true)) {
```

Comment:
could save some duplicated code by putting this into a method and having this and the scheduled runnable call it if ! isEmpty